### ISSUE TITLE
Constant boolean-valued defined functions get unfolded

### ISSUE THREAD

**Issue by lukemaurer** [GaloisInc/crucible#103]
It seems that the functions `\_ -> True` and `\_ -> False` get expanded by `applySymFn` even when they're not defined with unfolding criteria. 

```haskell
module Main where

import Data.Parameterized.Context
import Data.Parameterized.Nonce
import Data.Parameterized.Some
import What4.BaseTypes
import What4.Expr.Builder
import What4.Interface

-- | From @e@, construct @(\_ -> e) e@ and print it. The application should
-- not be unfolded, yet it is unfolded if @e@ is either @True@ or @False@.
test :: IsSymExprBuilder sym
     => sym
     -> BaseTypeRepr tp
     -> SymExpr sym tp
     -> IO ()
test sym tp e = do
  let Right sbl = userSymbol "fun"
  fun <- inlineDefineFun sym sbl (Empty :> tp) $ \_ -> return e
  val <- applySymFn sym fun (Empty :> e)
  print (printSymExpr val)

data TestState a = TestState

main :: IO ()
main = do
  Some ng <- newIONonceGenerator
  sym <- newExprBuilder TestState ng
  
  test sym BaseIntegerRepr =<< intLit sym 42
    -- prints "let fun ?@2:i = 42 in apply fun 42"
  test sym BaseBoolRepr (truePred sym)
    -- prints "true"
```

**Comment by robdockins** [comment]
@lukemaurer, is this bug still impeding work you're doing?

**Comment by lukemaurer** [comment]
I don't think so. Not that I'm currently doing, anyway.

**Comment by robdockins** [comment]
I think this bug has now been fixed, but I haven't yet verified that the example program above now has the expected behavior.

**Comment by robdockins** [comment]
OK, this is not yet fixed in master, but I know exactly what is causing it, and it will be fixed when #176 is merged.  We can cherry-pick a fix sooner, if necessary.

**Comment by robdockins** [comment]
Fixed in 671922140
